File: source/src/main/java/org/rss/tools/mpl/global/ExceptionHandler.java
package org.rss.tools.mpl.global;

import java.util.Arrays;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.AfterThrowing;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.rss.tools.mpl.validation.AppException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

@Aspect
public class ExceptionHandler {
	
	private static final Logger LOGGER = LoggerFactory.getLogger(ExceptionHandler.class);

	@Pointcut("call(* org.rss.tools.mpl.parsing..*(..))")
	public void parsePhase() {};
	
	@Pointcut("within(org.rss.tools.mpl.Main)")
	public void fromMain() {};
	
	@Pointcut("execution(* org.rss.tools.mpl.Main.main(*))")
	public void mainCall() {};
	
	// --------------------
	
	@Around("mainCall()")
	public Object globalHandler(ProceedingJoinPoint pjp) throws Throwable {
		try {
			return pjp.proceed();
		} catch (HandledException e) {
			LOGGER.debug("Exception already handled");
			return null;
		} catch (Exception e) {
			LOGGER.error("------------------------------------------------------");
			LOGGER.error("An unexpected error has ocurred. ", e);
			return null;
		} 
	}
	
	// ------------------------- Parsing ----------------------------
	
	@AfterThrowing(value = "fromMain() && parsePhase()", throwing = "e")
	public void throwOnParse(JoinPoint thisJoinPoint, Exception e) {
		LOGGER.debug("Error on Parse phase. ", e.getLocalizedMessage());
		handleAppException(thisJoinPoint, e);
	}

	// ------------------------- Render ----------------------------
	
	@AfterThrowing(value = "call(* org.rss.tools.mpl.render..*(..))", throwing = "e")
	public void afterThrowRender(JoinPoint thisJoinPoint, Exception e) {
		LOGGER.debug("Error on Render phase. ", e.getLocalizedMessage());
		handleAppException(thisJoinPoint, e);
	}
	
	// --------------------------------------------------------------
	private void handleAppException(JoinPoint thisJoinPoint, Exception e) {
		if (e instanceof AppException) {
			AppException pse = (AppException) e;
			if (LOGGER.isDebugEnabled()) {
				LOGGER.debug("Location: ", thisJoinPoint.getSignature());
				LOGGER.debug(Arrays.toString(thisJoinPoint.getArgs()));
			}

			printApplicationException(pse.getError().getMessage(), pse.getError().getHowFix(), pse.getParameters());
			throw new HandledException(e);
		}
	}
	
	private void printApplicationException(String error, String fixTip, Object... parameters) {
		LOGGER.error("------------------ An Error ocurred ------------------");
		LOGGER.error(String.format("%-15s: %s", "Error", error)); 
		if (parameters != null && parameters.length > 0) {
			LOGGER.error(String.format("%-15s: %s", "Parameters", Arrays.toString(parameters)));
		}
		LOGGER.error(String.format("%-15s: %s", "Possible fix", fixTip));
		LOGGER.error("------------------------------------------------------");
	}

	public static class HandledException extends RuntimeException {

		/**
		 * 
		 */
		private static final long serialVersionUID = 1L;

		public HandledException(Throwable cause) {
			super(cause);
		}
		
	}
}

File: source/src/main/kotlin/org/rss/tools/mpl/global/ExceptionHandler.kt
package org.rss.tools.mpl.global

import java.util.Arrays

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.AfterThrowing
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.rss.tools.mpl.validation.AppException
import org.slf4j.LoggerFactory

@Aspect
class ExceptionHandler {

    @Pointcut("call(* org.rss.tools.mpl.parsing..*(..))")
    fun parsePhase() {
    }

    @Pointcut("within(org.rss.tools.mpl.Main)")
    fun fromMain() {
    }

    @Pointcut("execution(* org.rss.tools.mpl.Main.main(*))")
    fun mainCall() {
    }

    // --------------------

    @Around("mainCall()")
    @Throws(Throwable::class)
    fun globalHandler(pjp: ProceedingJoinPoint): Any? {
        try {
            return pjp.proceed()
        } catch (e: HandledException) {
            LOGGER.debug("Exception already handled")
            return null
        } catch (e: Exception) {
            LOGGER.error("------------------------------------------------------")
            LOGGER.error("An unexpected error has ocurred. ", e)
            return null
        }

    }

    // ------------------------- Parsing ----------------------------

    @AfterThrowing(value = "fromMain() && parsePhase()", throwing = "e")
    fun throwOnParse(thisJoinPoint: JoinPoint, e: Exception) {
        LOGGER.debug("Error on Parse phase. ", e.localizedMessage)
        handleAppException(thisJoinPoint, e)
    }

    // ------------------------- Render ----------------------------

    @AfterThrowing(value = "call(* org.rss.tools.mpl.render..*(..))", throwing = "e")
    fun afterThrowRender(thisJoinPoint: JoinPoint, e: Exception) {
        LOGGER.debug("Error on Render phase. ", e.localizedMessage)
        handleAppException(thisJoinPoint, e)
    }

    // --------------------------------------------------------------
    private fun handleAppException(thisJoinPoint: JoinPoint, e: Exception) {
        if (e is AppException) {
            if (LOGGER.isDebugEnabled) {
                LOGGER.debug("Location: ", thisJoinPoint.signature)
                LOGGER.debug(Arrays.toString(thisJoinPoint.args))
            }

            printApplicationException(e.error!!.message, e.error!!.howFix, *e.parameters ?: emptyArray())
            throw HandledException(e)
        }
    }

    private fun printApplicationException(error: String, fixTip: String?, vararg parameters: Any) {
        with(LOGGER) {
            error("------------------ An Error ocurred ------------------")
            error(String.format("%-15s: %s", "Error", error))
            if (parameters.isNotEmpty()) {
                error(String.format("%-15s: %s", "Parameters", Arrays.toString(parameters)))
            }
            error(String.format("%-15s: %s", "Possible fix", fixTip))
            error("------------------------------------------------------")
        }
    }

    class HandledException(cause: Throwable) : RuntimeException(cause) {
        companion object {

            /**
             *
             */
            private const val serialVersionUID = 1L
        }

    }

    companion object {

        private val LOGGER = LoggerFactory.getLogger(ExceptionHandler::class.java)
    }
}
